### PR TITLE
fix(common): connections over Proxy not using correct `HTTP` protocol

### LINKED ISSUE #158: [Bug] Connection over Proxy not using correct protocol
### Plugin Name

@nxrocks/nx-spring-boot

### Plugin Version

6.3.0

### Nx Version

15.5.2

### Expected Behaviour

The proxy agent returned should be based on the protocol based on the URL and not on the Proxy. See also 
image

It seems that the Http(s)ProxyAgent needs to based on the Server URL and not on the Proxy URL.

I can also create an PR for this if you want ?

### Actual Behaviour

Currently it is trying to connect with the false protocol if the procotol from the proxy and server are not the same.

Here a short matrix based on the above image:
| Proxy | Server | Status      |
|-------|--------|-------------|
| HTTP  | HTTP   | working     |
| HTTPS | HTTP   | not working |
| HTTP  | HTTPS  | not working |
| HTTPS | HTTPS  | working     |

### Steps to reproduce the behaviour

1. Have a setup with a http proxy.
2. Generate a app which needs an connection.

### PR DESCRIPTION
**Pull request by Darkterror45**
fixes #158

### CODE REVIEW

File: packages/common/src/lib/core/utils/utils.ts [hunk 1]
```diff
@@ -11,7 +11,7 @@ export function getPackageLatestNpmVersion(pkg: string): string {
   }
 }
 
-export function getHttpProxyAgent(proxyUrl?: string): HttpProxyAgent | HttpsProxyAgent | undefined {
+export function getHttpProxyAgent(proxyUrl?: string, downloadUrl?: string): HttpProxyAgent | HttpsProxyAgent | undefined {
```
Comment (tinesoft) [inline packages/common/src/lib/core/utils/utils.ts:14]: can you please make `downloadUrl`  **not optional** (and so move it _before_  `proxyUrl` ?   as such:

```
export function getHttpProxyAgent(downloadUrl: string, proxyUrl?: string, ): HttpProxyAgent | HttpsProxyAgent | undefined {
```

File: packages/common/src/lib/core/utils/utils.ts [hunk 2]
```diff
@@ -34,19 +34,19 @@ export function getHttpProxyAgent(proxyUrl?: string): HttpProxyAgent | HttpsProx
     console.log(`The proxy server at '${proxy}' will be used.`);
   }
   
-  if (proxy?.startsWith('https')) {
+  if (downloadUrl?.startsWith('https')) {
```
Comment (tinesoft) [inline packages/common/src/lib/core/utils/utils.ts:37]: change the condition to:
```
  if (targetUrl.startsWith('https')  && !!proxyUrl) {
```

(_only set the proxy if needed (proxyUrl  is truthy)_)

File: packages/common/src/lib/core/utils/utils.ts [hunk 3]
```diff
@@ -34,19 +34,19 @@ export function getHttpProxyAgent(proxyUrl?: string): HttpProxyAgent | HttpsProx
...
     return new HttpsProxyAgent(proxyAgentOpts);
-  } else if(proxy?.startsWith('http')){
+  } else if(downloadUrl?.startsWith('http')){
```
Comment (tinesoft) [inline packages/common/src/lib/core/utils/utils.ts:39]: change the condition to:
```
  if (targetUrl.startsWith('http')  && !!proxyUrl) {
```

(_only set the proxy if needed (proxyUrl  is truthy)_)

File: packages/common/src/lib/core/utils/utils.ts [hunk 4]
```diff
@@ -34,19 +34,19 @@ export function getHttpProxyAgent(proxyUrl?: string): HttpProxyAgent | HttpsProx
...
     return new HttpProxyAgent(proxyAgentOpts);
   } else {
     return undefined
   }
 }
 
-export function getCommonHttpHeaders( pkgName: string, proxyUrl?: string){
+export function getCommonHttpHeaders( pkgName: string, proxyUrl?: string, downloadUrl?: string){
```
Comment (tinesoft) [inline packages/common/src/lib/core/utils/utils.ts:46]: can you please : 
- make `downloadUrl`  **not optional**,
- rename it into a more generic `targetUrl` 
- and so move it _before_  `proxyUrl`
- 
as such:

```
export function getCommonHttpHeaders( pkgName: string, targetUrl?: string, proxyUrl?: string){
```

File: packages/nx-micronaut/src/utils/micronaut-utils.ts [hunk 5]
```diff
@@ -77,7 +77,7 @@ export function isMicronautProject(project: ProjectConfiguration): boolean {
 
 
 export async function fetchMicronautFeatures(options:  NormalizedSchema): Promise<MicronautFeature[]> {
-    const response = await fetch(`${options.micronautLaunchUrl}/application-types/${options.projectType}/features`, getCommonHttpHeaders(NX_MICRONAUT_PKG, options.proxyUrl));
+    const response = await fetch(`${options.micronautLaunchUrl}/application-types/${options.projectType}/features`, getCommonHttpHeaders(NX_MICRONAUT_PKG, options.proxyUrl, options.micronautLaunchUrl));
```
Comment (tinesoft) [inline packages/nx-micronaut/src/utils/micronaut-utils.ts:80]: swap parameters of "getCommonHttpHeaders" based on previous refactoring

File: packages/nx-spring-boot/src/generators/project/lib/generate-boot-project.ts [hunk 6]
```diff
@@ -10,7 +10,7 @@ export async function generateBootProject(tree: Tree, options: NormalizedSchema)
 
     logger.info(`⬇️ Downloading Spring Boot project zip from : ${downloadUrl}...`);
 
-    const response = await fetch(downloadUrl, getCommonHttpHeaders(NX_SPRING_BOOT_PKG, options.proxyUrl));
+    const response = await fetch(downloadUrl, getCommonHttpHeaders(NX_SPRING_BOOT_PKG, options.proxyUrl, downloadUrl));
```
Comment (tinesoft) [inline packages/nx-spring-boot/src/generators/project/lib/generate-boot-project.ts:13]: swap parameters of "getCommonHttpHeaders" based on previous refactoring

File: packages/nx-quarkus/src/utils/quarkus-utils.ts [hunk 7]
```diff
@@ -65,7 +65,7 @@ export function isQuarkusProject(project: ProjectConfiguration): boolean {
 }
 
 export async function fetchQuarkusExtensions(options:  NormalizedSchema): Promise<QuarkusExtension[]> {
-    const response = await fetch(`${options.quarkusInitializerUrl}/api/extensions`, getCommonHttpHeaders(NX_QUARKUS_PKG, options.proxyUrl));
+    const response = await fetch(`${options.quarkusInitializerUrl}/api/extensions`, getCommonHttpHeaders(NX_QUARKUS_PKG, options.proxyUrl, options.quarkusInitializerUrl));
```
Comment (tinesoft) [inline packages/nx-quarkus/src/utils/quarkus-utils.ts:68]: swap parameters of "getCommonHttpHeaders" based on previous refactoring

**Review by tinesoft — CHANGES_REQUESTED** [review]
Thank you for the PR!

Can you address the changes suggested please?
I'll merge afterwards if validated

**Comment by tinesoft** [comment]
Hi @Darkterror45 

Did you have time to apply the requested changes?

If not, and if you don't mind, I'll take over and continue the PR (you'll still be the author of course ^^).

I'm asking because I'd like to release this before the end of the week if possible.

Thanks!

**Comment by Darkterror45** [comment]
Hi @tinesoft,

Yes, i have push the requested changes.

**Comment by tinesoft** [comment]
Thanks for the PR @Darkterror45 ! 🙏🏾

**Pull request merged**